core/src/avm2/globals/flash/display/graphics.rs

```diff
@@ -364,6 +365,365 @@ pub fn draw_rect<'gc>(
     Ok(Value::Undefined)
 }
 
+/// Implements `Graphics.drawRoundRect`.
+pub fn draw_round_rect<'gc>(
+    activation: &mut Activation<'_, 'gc, '_>,
+    this: Option<Object<'gc>>,
+    args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error> {
+    if let Some(this) = this.and_then(|t| t.as_display_object()) {
+        let x = args
+            .get(0)
+            .cloned()
+            .unwrap_or(Value::Undefined)
+            .coerce_to_number(activation)?;
+        let y = args
+            .get(1)
+            .cloned()
+            .unwrap_or(Value::Undefined)
+            .coerce_to_number(activation)?;
+        let width = args
+            .get(2)
+            .cloned()
+            .unwrap_or(Value::Undefined)
+            .coerce_to_number(activation)?;
+        let height = args
+            .get(3)
+            .cloned()
+            .unwrap_or(Value::Undefined)
+            .coerce_to_number(activation)?;
+        let mut ellipse_width = args
+            .get(4)
+            .cloned()
+            .unwrap_or(Value::Undefined)
+            .coerce_to_number(activation)?;
+        let mut ellipse_height = args
+            .get(5)
+            .cloned()
+            .unwrap_or(Value::Number(f64::NAN))
+            .coerce_to_number(activation)?;
+
+        if let Some(mut draw) = this.as_drawing(activation.context.gc_context) {
+            if ellipse_height.is_nan() {
+                ellipse_height = ellipse_width;
+            }
+
+            //Clamp the ellipse sizes to the size of the rectangle.
+            if ellipse_width > width {
+                ellipse_width = width;
+            }
+
+            if ellipse_height > height {
+                ellipse_height = height;
+            }
+
+            // We always draw ellipses in 45-degree chunks, so let's
+            // precalculate some common math. For efficiency, we'll work on a
+            // unit circle at the origin and then translate and scale to get
+            // four ellipse pieces on our roundrect. This lets us not have to
+            // care about scaling our constants by a potentially varying
+            // radius.
+            //
```

Comment:
If you took the math from somewhere, could you please link a source? I'm assuming you didn't derive it all by hand :D 

I found some other sources (via StackOverflow) that appear to skip trigonometry completely and appear to produce a single constant (plus scaling for ellipses): http://www.whizkidtech.redprince.net/bezier/circle/kappa/

Reply:
...also took me a while to notice the main difficulty is that FP need to them be quadratic :(